File: src/Sales/BrewUp.Sales.Entities/Entities/SalesOrderRow.cs
using BrewUp.Sales.SharedKernel.CustomTypes;
using BrewUp.Shared.ReadModel;

namespace BrewUp.Sales.Entities.Entities;

public class SalesOrderRow : DtoBase
{
    public string ProductId { get; private set; } = string.Empty;
    public string ProductName { get; private set; } = string.Empty;
    
    public double Quantity { get; private set; }
    public string UnitOfMeasure { get; private set; } = string.Empty;
    
    public double Price { get; private set; }
    public string Currency { get; private set; } = string.Empty;
    
    protected SalesOrderRow() 
    { }
    
    internal static SalesOrderRow Create(ProductId productId, ProductName productName, Quantity quantity, Price price)
    {
        return new SalesOrderRow(productId, productName, quantity, price);
    }

    private SalesOrderRow(ProductId productId, ProductName productName, Quantity quantity, Price price)
    {
        // _productId = productId;
        // _productName = productName;
        //
        // _quantity = quantity;
        // _price = price;
    }
}

File: src/Sales.FSharp/BrewUp.Sales.Entities.FSharp/Entities/SalesOrderRow.fs
namespace BrewUp.Sales.Entities.FSharp.Entities

open System
open BrewUp.Shared.FSharp.Domain
open BrewUp.Shared.FSharp.ExternalContracts
open BrewUp.Sales.SharedKernel.FSharp.CustomTypes

type SalesOrderRow = {
    Id: string
    SalesOrderId: string
    ProductId: string
    ProductName: string
    Quantity: float
    UnitOfMeasure: string
    Price: decimal
    Currency: string
}

module SalesOrderRow =
    let create (salesOrderId: SalesOrderId) (productId: ProductId) (productName: ProductName) 
               (quantity: ProductQuantity) (price: ProductPrice) =
        {
            Id = Guid.NewGuid().ToString()
            SalesOrderId = salesOrderId.Value
            ProductId = productId.Value
            ProductName = productName.Value
            Quantity = quantity.Quantity
            UnitOfMeasure = quantity.UnitOfMeasure
            Price = price.Price
            Currency = price.Currency
        }